qt/widgets/instrumentview/src/InstrumentRenderer.cpp

```diff
@@ -0,0 +1,439 @@
+
+namespace {
+size_t decodePickColorRGB(unsigned char r, unsigned char g, unsigned char b) {
+  unsigned int index = r;
+  index *= 256;
+  index += g;
+  index *= 256;
+  index += b - 1;
+  return index;
+}
+
+void updateVisited(std::vector<bool> &visited,
+                   const std::vector<size_t> &components) {
+  for (auto component : components)
+    visited[component] = true;
+}
+} // namespace
+
+InstrumentRenderer::InstrumentRenderer(const InstrumentActor &actor)
+    : m_actor(actor) {
+
+  m_displayListId[0] = 0;
+  m_displayListId[1] = 0;
+  m_useDisplayList[0] = false;
+  m_useDisplayList[1] = false;
+
+  const auto &componentInfo = actor.componentInfo();
+  size_t numTextures = 0;
+  for (size_t i = componentInfo.root(); !componentInfo.isDetector(i); --i) {
+    auto type = componentInfo.componentType(i);
+    if (type == Mantid::Beamline::ComponentType::Rectangular ||
+        type == Mantid::Beamline::ComponentType::OutlineComposite) {
+      m_textureIndices.push_back(i);
+      m_reverseTextureIndexMap[i] = numTextures;
+      numTextures++;
+    }
+  }
+
+  if (numTextures > 0) {
+    m_textureIDs.resize(numTextures, 0);
+    colorTextures.resize(numTextures);
+    pickTextures.resize(numTextures);
+  }
+}
+
+InstrumentRenderer::~InstrumentRenderer() {
+  for (size_t i = 0; i < 2; ++i) {
+    if (m_displayListId[i] != 0) {
+      glDeleteLists(m_displayListId[i], 1);
+    }
+  }
+}
+
+void InstrumentRenderer::renderInstrument(const std::vector<bool> &visibleComps,
+                                          bool showGuides, bool picking) {
+  if (std::none_of(visibleComps.cbegin(), visibleComps.cend(),
+                   [](bool visible) { return visible; }))
+    return;
+
+  OpenGLError::check("InstrumentActor::draw()");
+  size_t i = picking ? 1 : 0;
+  if (m_useDisplayList[i]) {
+    glCallList(m_displayListId[i]);
+  } else {
+    m_displayListId[i] = glGenLists(1);
+    m_useDisplayList[i] = true;
+    glNewList(m_displayListId[i],
+              GL_COMPILE); // Construct display list for object representation
+    draw(visibleComps, showGuides, picking);
+    glEndList();
+    if (glGetError() == GL_OUT_OF_MEMORY) // Throw an exception
+      throw Mantid::Kernel::Exception::OpenGLError(
+          "OpenGL: Out of video memory");
+    glCallList(m_displayListId[i]);
+  }
+  OpenGLError::check("InstrumentActor::draw()");
+}
+
+void InstrumentRenderer::draw(const std::vector<bool> &visibleComps,
+                              bool showGuides, bool picking) {
+  const auto &compInfo = m_actor.componentInfo();
+  std::vector<bool> visited(compInfo.size(), false);
+
+  for (size_t i = compInfo.root(); i != std::numeric_limits<size_t>::max();
+       --i) {
+    auto type = compInfo.componentType(i);
+    if (type == Mantid::Beamline::ComponentType::Infinite)
+      continue;
+
+    if (type == Mantid::Beamline::ComponentType::Rectangular) {
+      visited[i] = true;
+      updateVisited(visited, compInfo.children(i));
```

Comment:
`compInfo.children(i)` doesn't return the list of all the detectors but a list of rows (or columns). For LOQ it returns 128 numbers but it has 128 x 128 detectors. As a result both the texture and the individual detectors are drawn.